File: src/npl_mcp/npl/layout.py
"""
Layout strategies for formatting NPL components.

Provides different organization strategies for the output:
- YAML_ORDER: Preserve YAML definition order
- CLASSIC: Category-based organization
- GROUPED: Group by section type
"""

from enum import Enum
from typing import List, Dict, Any

from .resolver import ResolvedComponent


class LayoutStrategy(Enum):
    """Available layout strategies for output formatting."""
    YAML_ORDER = "yaml-order"      # Preserve YAML definition order
    CLASSIC = "classic"            # Category-based organization
    GROUPED = "grouped"            # Group by type/labels


class NPLLayoutEngine:
    """Formats resolved components into markdown output."""

    def __init__(self, strategy: LayoutStrategy = LayoutStrategy.YAML_ORDER):
        """Initialize with layout strategy.

        Args:
            strategy: The layout strategy to use
        """
        self.strategy = strategy

    def format(self, components: List[ResolvedComponent]) -> str:
        """Format components into markdown string.

        Args:
            components: List of resolved components to format

        Returns:
            Markdown formatted string with all components
        """
        if not components:
            return ""

        if self.strategy == LayoutStrategy.YAML_ORDER:
            return self._format_yaml_order(components)
        elif self.strategy == LayoutStrategy.CLASSIC:
            return self._format_classic(components)
        elif self.strategy == LayoutStrategy.GROUPED:
            return self._format_grouped(components)
        else:
            # Default to YAML order
            return self._format_yaml_order(components)

    def _format_yaml_order(self, components: List[ResolvedComponent]) -> str:
        """Format components in their original YAML order.

        Args:
            components: Components to format

        Returns:
            Markdown string
        """
        parts = []
        for comp in components:
            parts.append(self.format_component(comp))
        return "\n\n".join(parts)

    def _format_classic(self, components: List[ResolvedComponent]) -> str:
        """Format components organized by category/labels.

        Args:
            components: Components to format

        Returns:
            Markdown string
        """
        # Group by first label (if any), otherwise "Uncategorized"
        categories: Dict[str, List[ResolvedComponent]] = {}

        for comp in components:
            category = comp.labels[0] if comp.labels else "uncategorized"
            if category not in categories:
                categories[category] = []
            categories[category].append(comp)

        parts = []
        for category, comps in sorted(categories.items()):
            parts.append(f"## {category.title()}")
            for comp in comps:
                parts.append(self.format_component(comp))

        return "\n\n".join(parts)

    def _format_grouped(self, components: List[ResolvedComponent]) -> str:
        """Format components grouped by section type.

        Args:
            components: Components to format

        Returns:
            Markdown string
        """
        # Group by section
        groups: Dict[str, List[ResolvedComponent]] = {}

        for comp in components:
            section_name = comp.section.value
            if section_name not in groups:
                groups[section_name] = []
            groups[section_name].append(comp)

        parts = []
        for section_name, comps in groups.items():
            parts.append(f"## {section_name.replace('-', ' ').title()}")
            for comp in comps:
                parts.append(self.format_component(comp))

        return "\n\n".join(parts)

    def format_component(self, component: ResolvedComponent) -> str:
        """Format a single component to markdown.

        Args:
            component: The component to format

        Returns:
            Markdown string for this component
        """
        parts = []

        # Header with name and slug
        # Always include slug to enable lookup by slug in the output
        if component.slug:
            parts.append(f"### {component.name} (`{component.slug}`)")
        else:
            parts.append(f"### {component.name}")

        # Brief
        if component.brief:
            parts.append(f"*{component.brief}*")

        # Description
        if component.description:
            parts.append(component.description)

        # Syntax
        if component.syntax:
            parts.append("**Syntax:**")
            for syn in component.syntax:
                syn_text = syn.get('syntax', syn.get('name', ''))
                parts.append(f"- `{syn_text}`")

        # Examples
        if component.examples:
            parts.append("**Examples:**")
            parts.append(self.format_examples(component.examples))

        # Labels
        if component.labels:
            labels_str = ", ".join(f"`{l}`" for l in component.labels)
            parts.append(f"*Labels: {labels_str}*")

        # Requirements
        if component.require:
            requires_str = ", ".join(f"`{r}`" for r in component.require)
            parts.append(f"*Requires: {requires_str}*")

        return "\n\n".join(parts)

    def format_examples(self, examples: List[Dict[str, Any]]) -> str:
        """Format component examples to markdown.

        Args:
            examples: List of example dictionaries

        Returns:
            Markdown string with formatted examples
        """
        if not examples:
            return ""

        parts = []
        for example in examples:
            name = example.get('name', 'Example')
            brief = example.get('brief', '')
            priority = example.get('priority', 0)

            if brief:
                parts.append(f"- **{name}**: {brief} (priority {priority})")
            else:
                parts.append(f"- **{name}** (priority {priority})")

        return "\n".join(parts)

File: backend/lib/noizu_prompt_lingua/npl/layout.ex
defmodule NoizuPromptLingua.NPL.Layout do
  @moduledoc false

  alias NoizuPromptLingua.NPL.Resolver.ResolvedComponent

  @type strategy :: :yaml_order | :classic | :grouped

  @spec format([ResolvedComponent.t()], strategy()) :: String.t()
  def format([], _strategy), do: ""
  def format(components, :yaml_order), do: format_yaml_order(components)
  def format(components, :classic), do: format_classic(components)
  def format(components, :grouped), do: format_grouped(components)
  def format(components, _), do: format_yaml_order(components)

  defp format_yaml_order(components) do
    components
    |> Enum.map(&format_component/1)
    |> Enum.join("\n\n")
  end

  defp format_classic(components) do
    components
    |> Enum.group_by(fn comp ->
      case comp.labels do
        [first | _] -> first
        [] -> "uncategorized"
      end
    end)
    |> Enum.sort_by(fn {cat, _} -> cat end)
    |> Enum.map(fn {category, comps} ->
      header = "## #{title_case(category)}"
      body = Enum.map(comps, &format_component/1) |> Enum.join("\n\n")
      "#{header}\n\n#{body}"
    end)
    |> Enum.join("\n\n")
  end

  defp format_grouped(components) do
    components
    |> Enum.group_by(& &1.section)
    |> Enum.map(fn {section_name, comps} ->
      title = section_name |> String.replace("-", " ") |> title_case()
      header = "## #{title}"
      body = Enum.map(comps, &format_component/1) |> Enum.join("\n\n")
      "#{header}\n\n#{body}"
    end)
    |> Enum.join("\n\n")
  end

  defp format_component(%ResolvedComponent{} = comp) do
    parts = []

    parts =
      if comp.slug != "" do
        ["### #{comp.name} (`#{comp.slug}`)" | parts]
      else
        ["### #{comp.name}" | parts]
      end

    parts = if comp.brief != "", do: ["*#{comp.brief}*" | parts], else: parts
    parts = if comp.description != "", do: [comp.description | parts], else: parts

    parts =
      if comp.syntax != [] do
        syntax_lines =
          Enum.map(comp.syntax, fn syn ->
            text = Map.get(syn, "syntax", Map.get(syn, "name", ""))
            "- `#{text}`"
          end)

        ["**Syntax:**\n#{Enum.join(syntax_lines, "\n")}" | parts]
      else
        parts
      end

    parts =
      if comp.examples != [] do
        examples_text = format_examples(comp.examples)
        ["**Examples:**\n#{examples_text}" | parts]
      else
        parts
      end

    parts =
      if comp.labels != [] do
        labels_str = Enum.map(comp.labels, &"`#{&1}`") |> Enum.join(", ")
        ["*Labels: #{labels_str}*" | parts]
      else
        parts
      end

    parts =
      if comp.require != [] do
        requires_str = Enum.map(comp.require, &"`#{&1}`") |> Enum.join(", ")
        ["*Requires: #{requires_str}*" | parts]
      else
        parts
      end

    parts
    |> Enum.reverse()
    |> Enum.join("\n\n")
  end

  defp format_examples(examples) do
    Enum.map(examples, fn ex ->
      name = Map.get(ex, "name", "Example")
      brief = Map.get(ex, "brief", "")
      priority = Map.get(ex, "priority", 0)

      if brief != "" do
        "- **#{name}**: #{brief} (priority #{priority})"
      else
        "- **#{name}** (priority #{priority})"
      end
    end)
    |> Enum.join("\n")
  end

  defp title_case(str) do
    str
    |> String.split(~r/[\s_-]+/)
    |> Enum.map(fn word ->
      case String.split_at(word, 1) do
        {first, rest} -> String.upcase(first) <> rest
      end
    end)
    |> Enum.join(" ")
  end
end
